app/Http/Controllers/Auth/AuthController.php

```diff
@@ -267,14 +183,145 @@ public function emailChangeStep3($token)
         $user->save();
         DB::commit();
 
-        return self::json($user->toArrayCamelize(), 200);
+        return self::json($user->toArray(), 200);
+    }
+
+    /**
+     * Log the user out of the application.
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function check(Request $request)
+    {
+        return self::json(Auth::user()->toArray(), 200);
+    }
+
+    /**
+     * Log the user out of the application.
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function login(Request $request)
+    {
+        if (Auth::guard()->attempt(
+            $request->only('email', 'password'),
+            $request->filled('remember')
+        )) {
+            Auth::user()->generateApiKey();
+
+            return self::json(Auth::user()->load('AdserverWallet')->toArray(), 200);
+        }
+
+        return self::json([], 400);
+    }
+
+    /**
+     * Log the user out of the application.
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function logout(Request $request)
+    {
+        Auth::user()->clearApiKey();
+
+        return self::json([], 204);
+    }
+
+    /**
+     * Start password recovery process - generate and send email.
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function recovery(Request $request)
+    {
+        Validator::make($request->all(), ['email' => 'required|email', 'uri' => 'required'])->validate();
+        $user = User::where('email', $request->input('email'))->first();
+        if (empty($user)) {
+            return self::json([], 204);
+        }
+        DB::beginTransaction();
+        if (!Token::canGenerate($user->id, 'password-recovery', $this->password_recovery_resend_limit)) {
+            return self::json([], 204);
+        }
+        Mail::to($user)->queue(new AuthRecovery(
+            Token::generate('password-recovery', $this->password_recovery_token_time, $user->id),
+            $request->input('uri')
+        ));
+        DB::commit();
+
+        return self::json([], 204);
+    }
+
+    /**
+     * Tests and extends user password recovery token.
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function recoveryTokenExtend($token)
+    {
+        if (Token::extend($token, $this->password_recovery_token_time, null, 'password-recovery')) {
+            return self::json([], 204);
+        }
+
+        return self::json([], 422, ['message' => 'Password recovery token is invalid']);
     }
 
-    public function read(Request $request, $user_id)
+    public function updateSelf(Request $request)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this function 6, down to the maximum allowed 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1142)